File: flashfill/flashfill.py
#Reference: http://research.microsoft.com/en-us/um/people/sumitg/pubs/popl11-synthesis.pdf
import re
import itertools

class ConstStr:
    def __init__(self, s):
	self.s = s

class CPos:
    def __init__(self, p):
	self.p = p

class Pos:
    def __init__(self, r1, r2, c):
	self.r1 = r1
	self.r2 = r2
	self.c = c

class Substr:
    def __init__(self, v_i, y1, y2):
	self.v_i = v_i
	self.y1 = y1
	self.y2 = y2

def TokenSequence(s):
    TS = ""
    for char in s:
	if char >= 'A' and char <= 'Z':
	    TS += "U"
	elif char >= 'a' and char <= 'z':
	    TS += "L"
	elif char == ' ':
	    TS += "S"
    if len(s) == 0:
	return "E"
    return 'TS('+ ''.join(ch for ch, _ in itertools.groupby(TS)) + ')'

def ConcatenateTokenSequence(a, b):
    return "TS(" + a.split("TS(")[1].split(")")[0] + b.split("TS(")[1].split(")")[0] + ")"

def MatchTokenSequence(T, s):
    regex = ''

    if 'TS' in T:
	T = T.split("TS(")[1].split(")")[0]
    else:
	return regex

    for token in T:
	if token == 'U':
	    regex += '[A-Z]+'
        elif token == 'L':
	    regex += '[a-z]+'
	elif token == 'S':
	    regex += '\ +'
    return regex

def IParts(T, s):
    return T

def generateRegex(T, s):
    if T == "E":
	return ["E"]
    else:
        tokens = T.split("TS(")[1].split(")")[0] 
	regex = ""
	for token in tokens:
	    regex += IParts(token, s)
	regex = "TS(" + regex + ")"
        return [regex]

def generatePosition(s, k):
    #s - Input column
    #k - Start index of substring in the input column
    print "8) In generatePosition for", s, "and pos", k
 
    result = set([CPos(k), CPos(-(len(s) - k))])
	
    if k == 0:
	print "9)", "''", TokenSequence('')

    r1 = list() 
    k1_seq = list()
    for k1 in range(0, k):
	string = s[k1:k]
	token_sequence = TokenSequence(string)
	print "9)", string, token_sequence
        r1.append(token_sequence)
	k1_seq.append(k1)
    print

    r2 = list()
    k2_seq = list()
    for k2 in range(k+1, len(s)+1):
	string = s[k:k2]
        token_sequence = TokenSequence(string)
	print "10)", string, token_sequence
	r2.append(token_sequence)
	k2_seq.append(k2)
    print

    for a in r1:
	for b in r2:
	    r12 = ConcatenateTokenSequence(a, b)
            print "11) r12=", r12
	    regex = MatchTokenSequence(r12, s)
	
	    matches = list()
            for i in range(0, len(s)+1):
		for j in range(i+1, len(s)+1):
		    substring = s[i:j+1]
		    match = re.findall(regex, substring)
		    if len(match) == 1:
		         matches.append(match[0])
	    matches = list(set(matches)) 
 	    print "12) Matches", matches
 
            c_dash = len(matches)
	    print "13) c' = ", c_dash

	    k1 = k1_seq[r1.index(a)]
	    k2 = k2_seq[r2.index(b)] - 1

	    substring = s[k1:k2+1]
	    print "14) s[k1:k2]:", substring
     
            c = matches.index(substring)
	    print "15) c = ", c

            r1_dash = generateRegex(a, s)
	    r2_dash = generateRegex(b, s)
     
            print "16) r1_dash: ", r1_dash
	    print "17) r2_dash: ", r2_dash
        
	    positions = list()
            for reg1 in r1_dash:
		for reg2 in r2_dash:
		    for regc in [c, -(c_dash - c + 1)]:
                        p = Pos(reg1, reg2, regc)
			positions.append(p)

	    result = result.union(set(positions))
            print "result", list(result)
	    print
    return result
        
def generateSubstring(sigma, s):
    #sigma - List of input columns
    #s - substring
    print "5) In generateSubstring for", s
    result = set()

    matches = list()
    for i in range(0, len(sigma)):
        for match in re.finditer(re.escape(s), sigma[i]):
            matches.append((i, match.start()))
   
    for i,k in matches:
         print "6) Matches found in column", i, '=', sigma[i]
         print "7) Match: At pos", k, '=', str(sigma[i][k:len(s)+k])
         print

         y1 = generatePosition(sigma[i], k)
	 y2 = generatePosition(sigma[i], k + len(s))
 
         substrings = list()
         for reg_y1 in y1:
	     for reg_y2 in y2:
	         substrings.append(Substr(sigma[i], reg_y1, reg_y2))

         result = result.union(set(substrings))
    return result
   
def generateStr(sigma, s):
    #sigma - List of input columns
    #s - Output string
    nodes = set([i for i in xrange(len(s))])
    source_node = set([0])
    target_node = set([len(s)])

    edges = []
    for j in range(1, len(s)+1):
        for i in range(0, j):
	   edges.append((i, j)) 
    print "3) Edges:", edges
    print

    W = {}
    for edge in edges:
        i = edge[0]
        j = edge[1]
        substring = s[i:j]
        print "4) Calling generateSubstring on", substring
        W[(i,j)] = set([substring]).union(generateSubstring(sigma, substring))
	print "16) W: ", W 
    
    generateLoop(sigma, s, W)            
    #W = {edge:set([ConstStr(s[edge[0]:edge[1]-1])]).union(set(generateSubstring(sigma, s[edge[0]:edge[1]-1]))) for edge in edges}

def generateLoop(sigma, s, W):
    W_dash = W

    e1_dash = list()
    e2_dash = list()

    for k1 in range(0, len(s)+1):
        for k2 in range(0, len(s)+1):
            for k3 in range(0, len(s)+1):
                e1 = generateStr(sigma, s[k1:k2+1])
                print "17) e1 =", e1

def generateStringProgram(S):
    # S - Set of (sigma, s) pairs for input, output examples
    T = set()
    for each_input, each_output in S:
      each_sigma = each_input.split(',')
      print "1) Input:", each_sigma
      print "2) Output:", each_output
      print
      generateStr(each_sigma, each_output)

S = set([("Foo Bar","FB"), ("Text, String", "T")])
generateStringProgram(S)

File: flashfill/flashfill.rs
/* 
  Reference: http://research.microsoft.com/en-us/um/people/sumitg/pubs/popl11-synthesis.pdf
*/

// The program takes user-defined input-output examples to synthesize a set a programs consistent with the examples.
// The input of an example can span across multiple columns in a spreadsheet.
// The output of an example is expressed in one spreadsheet column.
// The contents of any spreadsheet cell is treated as a string.

// For simplifying the implementation, I've used alternative data structures in many places.
// For example, The paper represents the input of an example as a tuple of spreadsheet columns.
// I'm using a Vec of spreadsheet columns instead. Column contents are represented as strings as mentioned before. 
type SpreadsheetColumn = String;
type ExampleInput      = Vec<SpreadsheetColumn>;

// Assuming a set of tokens to get Example 3 from the paper to work
// TODO - Figure out how to generate a comprehensive set of Tokens for a given set of input-output examples
enum SpecialToken { SlashTok }

#[derive(Clone)]
enum Token        { SpecialToken }
type RegularExpr  = Vec<Token>;

// TOOD - Fix type of int
type int = usize;
enum Position    { CPos {k:int},
                   Pos  {r1:RegularExpr, r2:RegularExpr, c:Vec<int>} }

// TODO - Need to define struct Loop
// TODO - Redefine Concatenate to use DAGS to represent atomic expressions
enum AtomicExpr  { SubStr   {v:String, p1:Vec<Position>, p2:Vec<Position>},
                   ConstStr {s:String} }
type Concatenate = Vec<AtomicExpr>;
type DAG         = Concatenate;

struct Match     { v:String, r:RegularExpr, k:int }
struct NotMatch  { v:String, r:RegularExpr, k:int }
enum Predicate   { Match    {v:String, r:RegularExpr, k:int},
                   NotMatch {v:String, r:RegularExpr, k:int} }

type Conjuct = Vec<Predicate>;
type Bool    = Vec<Conjuct>;

struct Switch { bool_vec:Vec<Bool>, e_vec:Vec<DAG> } 
type StringExpr = Switch;

type ExampleInputSet = Vec<ExampleInput>;
struct Traces { b_vec:Vec<ExampleInputSet>, e_vec:Vec<DAG> }

fn util_vec_diff<T> (a: Vec<T>, b:Vec<T>) -> Vec<T> {
    panic!();
}

fn util_bool_expr_sort (b_vec:Vec<ExampleInputSet>, e_vec:Vec<DAG>) -> Traces {
    panic!();
}

fn util_compatibility (e_vec1:Vec<DAG>, e_vec2:Vec<DAG>) -> bool {
    panic!();
}

fn util_matching_tokens (s:String) -> Vec<RegularExpr> {
    panic!();
}

fn util_cth_match (substr:String, r:RegularExpr, s:String) -> (int, int) {
    panic!();
}

fn util_substr_pos (s:String, substr:String) -> int {
    panic!();
}
fn iParts (tok:Token, s:String) -> Token {
    panic!();
}

fn generate_string_program (example_inputs:Vec<ExampleInput>, example_outputs:Vec<SpreadsheetColumn>) -> StringExpr {
    let mut b_vec_init:  Vec<ExampleInputSet> = Vec::new();
    let mut e_vec_init:  Vec<DAG>             = Vec::new();
    let mut traces:      Traces               = Traces { b_vec:b_vec_init, e_vec:e_vec_init };
    
    for i in 0..example_inputs.len() {
        traces.b_vec.push(vec![example_inputs[i].clone()]);
        let dag = generate_str(example_inputs[i].clone(), example_outputs[i].clone());
        traces.e_vec.push(dag);
    }
    traces = generate_partition(example_inputs, example_outputs);

    let mut bool_classifiers: Vec<Bool> = Vec::new();
    for i in 0..traces.b_vec.len() {
        bool_classifiers.push(generate_bool_classifier(traces.b_vec[i].clone(), traces.b_vec[i].clone()));
    }
    traces = util_bool_expr_sort(traces.b_vec, traces.e_vec);

    let switch: Switch = Switch { bool_vec:bool_classifiers, e_vec:traces.e_vec };
    switch
}

fn generate_partition (example_inputs:Vec<ExampleInput>, example_outputs:Vec<SpreadsheetColumn>) -> Traces {
    panic!();
}

fn generate_bool_classifier (example_inputs1:Vec<ExampleInput>, example_inputs2:Vec<ExampleInput>) -> Bool {
    panic!();
}

fn generate_str (example_input:ExampleInput, example_output:SpreadsheetColumn) -> DAG {
    panic!();
}

// TODO Add parameter W
fn generate_loop (example_input:ExampleInput, example_output:SpreadsheetColumn) {
    panic!();
}

fn generate_substring (example_input:ExampleInput, s:String) -> Vec<AtomicExpr> {
    let mut result: Vec<AtomicExpr> = Vec::new();
    for i in 0..example_input.len() {
        let example_input_col = example_input[i].clone();
        let k  = util_substr_pos(s.clone(), example_input_col.clone());
        let y1 = generate_position(example_input_col.clone(), k);
        let y2 = generate_position(example_input_col.clone(), (k + s.len()));
        let substring: AtomicExpr = AtomicExpr::SubStr { v:example_input_col.clone(), p1:y1, p2:y2 };
    }
    result
}

fn generate_position (s:String, k:int) -> Vec<Position> {
    let s_len_as_int = s.len() as int;
    let mut result: Vec<Position> = Vec::new();
    let cpos_init_1: Position = Position::CPos { k:k };
    let cpos_init_2: Position = Position::CPos { k:(k - s_len_as_int) };
    result.push(cpos_init_1);
    result.push(cpos_init_2);

    for k1 in 0..k {
        for k2 in k..s_len_as_int {
            let substring1 : String = String::from(&s[k1 .. k]);
            let substring2 : String = String::from(&s[k .. k2]);
            let r1_set: Vec<RegularExpr> = util_matching_tokens(substring1.clone());
            let r2_set: Vec<RegularExpr> = util_matching_tokens(substring2.clone());
            for i in 0..substring1.len() {
                for j in 0..substring2.len() {
                    let mut r1 = r1_set[i].clone();
                    let mut r2 = r2_set[j].clone();
                    r1.extend(r2.clone());

                    let substring : String = String::from(&s[k1 .. k2]);
                    let (c, total) = util_cth_match(substring, r1.clone(), s.clone());

                    let regex1 = generate_regex(r1.clone(), s.clone());
                    let regex2 = generate_regex(r2.clone(), s.clone());

                    let pos: Position = Position::Pos { r1:regex1, r2:regex2, c:vec![c, ((c+1) - total)] };
                }
            }
        }
    }
    result
}

fn generate_regex (r:RegularExpr, s:String) -> RegularExpr {
    let mut regex: Vec<Token> = Vec::new();
    for i in 0..r.len() {
        regex.push(iParts(r[i].clone(), s.clone()));
    }
    regex
}
    
/*
type State = Vec<String>;
type StateSet = Vec<State>;

struct E {}

struct T {
  states : StateSet,
  exp : E,
}

struct ConstStr {
  s : String,
}

struct SubStr {
  v : String,
  p1 : usize,
  p2 : usize, 
}

struct CPos {
  k : usize,
}

enum Token { StartTok, EndTok, AlphaTok, SlashTok }

enum f { ConstStr, SubStr }

fn regex_of_str(s:String) -> Vec<Token>{
  let mut res: Vec<Token> = Vec::new();
  res.push(Token::StartTok);
  for c in s.chars() {
    match c {
      '\\' => res.push(Token::SlashTok),
      _ => res.push(Token::AlphaTok),
    }
  }
  res.push(Token::EndTok);
  res
}

fn generate_regex(r:Vec<Token>, s:String) -> Vec<Token> {r}

fn generate_position(s:String, k:usize){
  let mut result: Vec<CPos> = Vec::new();
  result.push(CPos{k:k});
  result.push(CPos{k:k-s.len()});
}

fn generate_substring(input_state:State, s:String) {
}

fn generate_str(input_state:State, output_string:String) -> E{
  let eta: Vec<usize> = (0..output_string.len()).collect();
  let eta_s: Vec<usize> = vec![0];
  let eta_t: Vec<usize> = vec![output_string.len()];
  let mut xi: Vec<(usize, usize)> = Vec::new();
  for j in 1..output_string.len() {
    for i in 0..j {
      xi.push((i,j));      
    }
  }
  let mut w: Vec<f> = Vec::new();
  for i in 1..xi.len() {
    let edge = xi[i];
    let ss = &output_string[edge.0..edge.0 + edge.1 - 1];
    let cs = ConstStr {s:ss.to_string()};
  }
  panic!()
}

fn generate_string_program(input_states:Vec<State>, output_strings:Vec<String>) {
  let mut t : Vec<T> = Vec::new();
  for i in 0..input_states.len() {
    let generated_strs = generate_str(input_states[i].clone(), output_strings[i].clone());
    let tt : T = T {states: vec![input_states[i].clone()], exp: generated_strs};
    t.push(tt);
  }
}

*/
fn main() {}
